rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+import collections
+import pkg_resources
+import socket
+import struct
+
+from .consts import SCPCommands, DataType
+from . import boot, consts
+from .scp_connection import SCPConnection
+from rig.utils.contexts import ContextMixin, Required
+
+
+class MachineController(ContextMixin):
+    """TODO
+
+    Attributes
+    ----------
+    """
+    def __init__(self, initial_host, n_tries=5, timeout=0.5, app_id=None,
+                 boot_data=None, struct_data=None):
+        """Create a new controller for a SpiNNaker machine.
+
+        Parameters
+        ----------
+        initial_host : string
+        n_tries : int
+        timeout : float
+        app_id : int or None
+        boot_data : bytes or None
+            Data to boot the machine with.  If None then the default boot file
+            provided with rig will be used.
+        struct_data : bytes or None
```

Comment:
Ah, so sv isn't read from the machine but rather the version pumped into the boot file is used? Also, isn't this a string (i.e. the contents of sark.struct, an ASCII formatted file)?